arrows/vxl/pixel_feature_extractor.cxx

```diff
@@ -98,18 +115,54 @@ ::check_sub_algorithm( vital::config_block_sptr config, std::string key )
   return true;
 }
 
+// ----------------------------------------------------------------------------
+vil_image_view< vxl_byte >
+pixel_feature_extractor::priv
+::generate_spatial_prior( kwiver::vital::image_container_sptr input_image )
+{
+  auto const image_data =
+    vxl::image_container::vital_to_vxl( input_image->get_image() );
+  // Return the previously-computed one if the size is the same
+  if( spatial_prior.ni() == image_data->ni() &&
+      spatial_prior.nj() == image_data->nj() )
+  {
+    return spatial_prior;
+  }
+
+  spatial_prior = vil_image_view< vxl_byte >( image_data->ni(),
+                                              image_data->nj(), 1 );
+
+  double const scale_factor =
+    static_cast< double >( std::numeric_limits< vxl_byte >::max() ) /
+    ( grid_length * grid_length - 1 );
+
+  for( unsigned i = 0; i < image_data->ni(); i++ )
+  {
+    auto const i_id =
+      static_cast< vxl_byte >( ( grid_length * i ) / image_data->ni() );
+    for( unsigned j = 0; j < image_data->nj(); j++ )
+    {
+      auto const j_id =
+        static_cast< vxl_byte >( ( grid_length * j ) / image_data->nj() );
+      auto const index = grid_length * j_id + i_id;
+      spatial_prior( i, j ) = static_cast< vxl_byte >( index * scale_factor );
+    }
+  }
+  return spatial_prior;
```

Comment:
Rather than trying to make lots and lots of individual comments on this:
```suggestion
  auto const ni = static_cast< unsigned >( input_image->width() );
  auto const nj = static_cast< unsigned >( input_image->height() );

  // Return the previously-computed one if the size is the same
  if( spatial_prior.ni() == ni && spatial_prior.nj() == nj )
  {
    return spatial_prior;
  }

  spatial_prior = vil_image_view< vxl_byte >( ni, nj, 1 );

  double const scale_factor =
    static_cast< double >( std::numeric_limits< vxl_byte >::max() ) /
    ( grid_length * grid_length - 1 );

  for( auto const i : kvr::iota( ni ) )
  {
    auto const i_id = ( grid_length * i ) / ni;
    for( auto const j : kvr::iota( nj ) )
    {
      auto const j_id = ( grid_length * j ) / nj;
      auto const index = grid_length * j_id + i_id;
      spatial_prior( i, j ) = static_cast< vxl_byte >( index * scale_factor );
    }
  }
  return spatial_prior;
```

Some comments:
- There doesn't seem to be any need to ever actually get the image, since only the dimensions are used. (And "caching" those makes the code shorter.)
- The early casts to byte for `i_id` and `j_id` don't seem necessary?

Reply:
Ahem: also...
```diff
diff --git a/arrows/vxl/pixel_feature_extractor.cxx b/arrows/vxl/pixel_feature_extractor.cxx
index 3546b16d8..3274d6c54 100644
--- a/arrows/vxl/pixel_feature_extractor.cxx
+++ b/arrows/vxl/pixel_feature_extractor.cxx
@@ -11,8 +11,11 @@
 #include "high_pass_filter.h"
 
 #include <arrows/vxl/image_container.h>
+
 #include <vital/config/config_block_io.h>
 
+#include <vital/range/iota.h>
+
 #include <vil/vil_clamp.h>
 #include <vil/vil_convert.h>
 #include <vil/vil_image_view.h>
@@ -23,6 +26,8 @@
 #include <limits>
 #include <type_traits>
 
+namespace kvr = kwiver::vital::range;
+
 namespace kwiver {
 
 namespace arrows {
```